warehouse/integrations/vulnerabilities/utils.py

```diff
@@ -0,0 +1,160 @@
+        .one()
+    )
+
+
+def _get_vuln_record(
+    request, vuln_report: vulnerabilities.VulnerabilityReportRequest, origin
+):
+    return (
+        request.db.query(VulnerabilityRecord)
+        .filter(VulnerabilityRecord.id == vuln_report.vulnerability_id)
+        .filter(VulnerabilityRecord.source == origin)
+        .options(orm.joinedload(VulnerabilityRecord.releases))
+        .one()
+    )
+
+
+def _delete_vuln_record(request, vuln_record: VulnerabilityRecord):
+    request.db.delete(vuln_record)
+
+
+def _add_vuln_record(request, vuln_record: VulnerabilityRecord):
+    request.db.add(vuln_record)
+
+
+def _analyze_vulnerability(request, vulnerability_report, origin, metrics):
+    try:
+        report = vulnerabilities.VulnerabilityReportRequest.from_api_request(
+            request=vulnerability_report
+        )
+    except vulnerabilities.InvalidVulnerabilityReportRequest as exc:
+        metrics.increment(
+            f"warehouse.vulnerabilities.error.{exc.reason}", tags=[f"origin:{origin}"]
+        )
+        raise
+
+    metrics.increment("warehouse.vulnerabilities.valid", tags=[f"origin:{origin}"])
+
+    try:
+        vulnerability_record = _get_vuln_record(request, report, origin)
+
+        if not report.versions:
+            # No versions indicates the vulnerability is no longer considered
+            # valid, so delete it.
+            _delete_vuln_record(request, vulnerability_record)
+            return
+
+    except NoResultFound:
+        if not report.versions:
+            return
+
+        vulnerability_record = VulnerabilityRecord(
+            id=report.vulnerability_id,
+            source=origin,
+            link=report.advisory_link,
+            aliases=report.aliases,
+        )
+        _add_vuln_record(request, vulnerability_record)
+
+    try:
+        project = _get_project(request, report)
+    except NoResultFound:
+        metrics.increment(
+            "warehouse.vulnerabilities.error.project_not_found",
+            tags=[f"origin:{origin}"],
+        )
+        raise
+
+    for version in report.versions:
+        try:
+            release = _get_release(request, project, version)
+        except NoResultFound:
+            metrics.increment(
+                "warehouse.vulnerabilities.error.release_not_found",
+                tags=[f"origin:{origin}"],
+            )
+            raise
+
+        if release not in vulnerability_record.releases:
+            vulnerability_record.releases.append(release)
+
+    # Unassociate any releases that no longer apply.
+    for release in list(vulnerability_record.releases):
+        if release.canonical_version not in report.versions:
+            vulnerability_record.releases.remove(release)
+
+
+def analyze_vulnerability(request, vulnerability_report, origin, metrics):
+    metrics.increment("warehouse.vulnerabilities.received", tags=[f"origin:{origin}"])
+    try:
+        _analyze_vulnerability(
```

Comment:
let's wrap this call in `with metrics.timed('warehouse.vulnerabilities.analysis', tags=[f"origin:{origin}"]):`

this will emit timing metrics so we can keep tabs on this programatically.

Reply:
done, good idea